lib/ipfs/Ipfs.ts

```diff
@@ -0,0 +1,163 @@
+
+    // An exmaple of multipart form data:
+    //
+    // --ABoundaryString
+    //
+    // Content of the first part.
+    // --ABoundaryString
+    // Content-Type: application/octet-stream
+    //
+    // Binary content of second part.
+    // --ABoundaryString--
+    const beginBoundary = Buffer.from(`--${multipartBoundaryString}\n`);
+    const firstPartContentType = Buffer.from(`Content-Type: application/octet-stream\n\n`);
+    const endBoundary = Buffer.from(`\n--${multipartBoundaryString}--`);
+    const requestBody = Buffer.concat([beginBoundary, firstPartContentType, content, endBoundary]);
+
+    const requestParameters = {
+      method: 'POST',
+      headers: { 'Content-Type': `multipart/form-data; boundary=${multipartBoundaryString}` },
+      body: requestBody
+    };
+
+    const addUrl = url.resolve(this.uri, '/api/v0/add'); // e.g. 'http://127.0.0.1:5001/api/v0/add'
+    const response = await this.fetch(addUrl, requestParameters);
+
+    if (response.status !== HttpStatus.OK) {
+      console.error(`IPFS write error response status: ${response.status}`);
+
+      if (response.body) {
+        const errorBody = await ReadableStream.readAll(response.body);
+        console.error(`IPFS write error body: ${errorBody}`);
+      }
+
+      throw new SidetreeError(IpfsErrorCode.IpfsFailedWritingContent, `Failed writing content of ${content.length} bytes.`);
+    }
+
+    const body = await ReadableStream.readAll(response.body);
+    const base58EncodedMultihashString = JSON.parse(body.toString()).Hash;
+
+    // Convert base58 to base64url multihash.
+    const multihashBuffer = multihashes.fromB58String(base58EncodedMultihashString);
+    const base64urlEncodedMultihash = base64url.encode(multihashBuffer);
+
+    console.log(`Wrote ${content.length} byte content as IPFS CID: ${base58EncodedMultihashString}, base64url ID: ${base64urlEncodedMultihash}`);
+    return base64urlEncodedMultihash;
+  }
+
+  public async read (base64urlEncodedMultihash: string, maxSizeInBytes: number): Promise<FetchResult> {
+    // Convert base64url to base58 multihash.
+    let base58EncodedMultihashString;
+    try {
+      const multihashBuffer = base64url.toBuffer(base64urlEncodedMultihash);
+      multihashes.validate(multihashBuffer);
+      base58EncodedMultihashString = multihashes.toB58String(multihashBuffer);
+    } catch {
+      return { code: FetchResultCode.InvalidHash };
+    }
+
+    const fetchContentPromise = this.fetchContent(base58EncodedMultihashString, maxSizeInBytes);
+    const fetchResult = await Timeout.timeout(fetchContentPromise, this.fetchTimeoutInSeconds * 1000);
+
+    // Mark content as `not found` if any error is thrown while fetching.
+    if (fetchResult instanceof Error) {
+      // Log appropriately based on error.
+      if (fetchResult instanceof SidetreeError &&
+          fetchResult.code === IpfsErrorCode.TimeoutPromiseTimedOut) {
+        console.log(`'${base64urlEncodedMultihash}' not found on IPFS.`);
```

Comment:
Discussion: Do we want to log both the base64 CID and the base58 CID? I think it will help make debugging easier since we won't need to convert them using some other tool, we can just copy and paste.

Reply:
Good suggestion... I tried to do it but only ended up doing it in happy path after a refactoring...